misc/Cgraph/utils-NeXT/open.1.0.NI.s/open.1
.TH OPEN 1L "Local Commands" "" \" -*- nroff -*-
.SH NAME
open, appopen, run, unhide \- open files and applications
.SH SYNOPSIS
.\" -activate and -makeKey are not yet documented, since they are the
.\" same as -unhide for now.
.B open
[\-a \fIapp\fR] [\-o \fIfile\fR] [\-p] [\-NXHost \fIhostname\fR]
[\-unhide] [\-nostat] [\-wait] [\-temp] [+\fIlinenum\fB]
[\fIfilename\fR]
\&.\|.\|.
.br
.B appopen
.I app
.br
.B run
.I app
.br
.B unhide
.I app
.br
.I app
[\fIoptions\fR]
.SH DESCRIPTION
The
.B open
command allows shell scripts or
.B Terminal
users to make requests to
NEXTSTEP applications. The applications may be launched or unhidden,
can be asked to open or print files. Besides the added options, it
should behave exactly like earlier (working) versions of NeXT's
.B open
command did. It does not require the ``Public Window Server'' switch (a
major security hole) in the
.B Preferences
application to be turned on to work correctly.

If no options are given at all or the special file name "\-" is used,
standard input is read. PostScript (.ps), Encapsulated PostScript
(.eps), and Rich Text Format (.rtf) files are recognized on standard
input and sent to the appropriate applications. All other files are
considered typeless, i.e. they are opened in either the named
application (see \fI\-a\fR) or the default application (normally
\fBEdit\fR).

Wherever the word \fIfile\fR is used, it also applies to wrappers and
ordinary directories.
.SS OPTIONS
.TP
.RI "\-a" " app"
All further requests are directed to the named application. The
application is launched if it is not running yet. If no
.I \-a
option is given, the Workspace is asked for the default application
for the files.
.TP
.RI "\-o" " file"
The application will be asked to open the named file. This option can
be used to switch back from print mode or to prefix a file name which
could be mistaken as an option. File names can also be escaped from
being mistaken as an option by prefixing them with `./' (this also
works with all other commands).
.TP
.RI "\-p"
The application will be asked to print the specified files. Print mode
will apply to all files given up to the next
.I "\-o"
option.
.TP
.RI "\-NXHost" " hostname"
All requests are directed to the Workspace and applications on the
remote host. The files to be accessed are looked for on the remote
host's file system. This option requires the ``Public Window Server''
switch on the remote host to be turned on, which is major security
hole, if you are connected to a network.
.TP
.RI "\-unhide"
The named application is unhidden. Most applications will also unhide
on a file opening request.
.TP
.RI "\-nostat"
The checks whether the next file to be opened or printed does not
exist will not be performed. In case of failure of a request to an
application, it is not possible to find out whether the file did not
exist or the application could not open it. It is necessary for
.IR \-NXHost ,
unless there is also a local file with the same name as the remote
file. This option also allows certain applications (such as
.BR Emacs )
to open files that do not exist yet.
.TP
.RI "\-wait"
.B open
will not continue until the next file given is saved for the first time.
.TP
.RI "\-temp"
The next file will be opened as a temporary file. This tells the
application it should remove the file when it is done. Despite it
belongs to the official API, only few applications (such as
.BR Preview )
honor this request.
.TP
.RI "+" "linenum"
In the next file, line number
.I linenum
will automatically be selected. This is currently only supported by
.B Edit
and
.BR Emacs .
.SS ALTERNATE NAMES
.br
.B open
can also be invoked by other names, where some defaults are
automatically set.
.TP
.BI "run" " app," " appopen" " app"
Behave as if it was invoked as
.BI "open -a" " app" \fR.
.TP
.BI "unhide" " app"
Identical to
.BI "open -a" " app" " -unhide" \fR.
.TP
.I "app"
If
.B open
is invoked with any other name, it behaves as if it were called as
.BI "open -a" " app" \fR.
.SH EXAMPLES
.TP
.B "open *.rtf"
opens all files ending in `.rtf' in the default application for this
file type (usually
.BR Edit ).
.TP
.B "open -p *.rtf"
prints all these files.
.TP
.B "open /tmp"
opens up a new file viewer showing the contents of the named
directory.
.TP
.B "appopen Yap Test.ps"
launches
.B Yap
(if it is not already running) and asks it to open
.BR Test.ps .
.TP
.B "Emacs *"
opens all files in
.B Emacs
(`Emacs' needs to be a link to `open').
.TP
.B "open -a Emacs -wait snd.2935"
asks
.B Emacs
to open the file.
.B open
will not continue unless the file has been saved. This is an
alternative to using
.B emacsclient
as your editor from within Unix programs.
.SH BUGS
The
.IR \-p ", " \-temp
and
.I +linenum
options are only supported by few applications.

.I \-NXHost
should imply
.IR \-nostat .
The current behaviour is compatible to the original
.BR open .

.I \-NXHost
requires ``Public Window Server'' (a major security hole) to be turned
on (this is
.B not
our fault).
.\" *HINT, HINT*

.SH AUTHORS
This version of
.B open
was written by Christian Limpach <chris@nice.ch>, with suggestions by
Felix Rauch <felix@nice.ch> and Marcel Waldvogel
<waldvogel@uptime.ch>. This manual page was written by Marcel
Waldvogel.


misc/Cgraph/utils-NeXT/open.1.0.NI.s/open.m

/* Author: Christian Limpach (chris@nice.ch) */

/*
 * /usr/local/lib/cvs/open/open.m,v 1.2 1994/05/09 21:37:42 chris Exp
 *
 * open.m,v
 * Revision 1.2  1994/05/09  21:37:42  chris
 * fixed NXHost which was broken because of [Application new]
 *
 * Revision 1.1  1994/05/05  22:32:05  chris
 * replaces /usr/bin/open and adds some nice features
 *
 */

#ifdef RCSID
static char rcsid[] = "open.m,v 1.2 1994/05/09 21:37:42 chris Exp";
#endif

#import "open.h"

@implementation AppInfo

- initFor:(const char *)openApp onHost:(const char *)hostName
{
  [super init];
  if((Port=NXPortFromName(openApp, hostName)) != PORT_NULL)
  {
    Name=str_copy(openApp);
    Host=str_copy(hostName);
  }
  return self;
}

- (port_t)port
{
  return Port;
}

- (BOOL)isEntryFor:(const char *)openApp:(const char *)hostName
{
  if(!strcmp(Name, openApp) &&
     (Host==hostName || (Host!=NULL && hostName!=NULL &&
			 !strcmp(Host, hostName))))
    return YES;
  else
    return NO;
}

@end

@implementation AppList

- (port_t)getPortFor:(const char *)anApp onHost:(const char *)aHost
{
  int index;
  
  for(index=0; index<[self count]; index++)
    if([[self objectAt:index] isEntryFor:anApp:aHost])
      break;
  if(index!=[self count])
    return [[self objectAt:index] port];
  else {
    id info;
    
    info=[[AppInfo alloc] initFor:anApp onHost:aHost];
    if ([info port]!=PORT_NULL)
      [self addObject:info];
    return [info port];
  }
}

@end

void main(int argc, char *argv[]) {
  port_t appPort;
  int ok, arg=1;
  char *host=NULL;
  const char *openApp=WORKSPACE, *appName;
  id appSpeaker = [[Speaker alloc] init];
  id appList = [[AppList alloc] init];
  int openMode=OPEN, fileMode=DISK, openedAFile=0;
  int oline=0, waitForFileChange=0, noStat=0;
  char *myArgv[argc];
  int i;

  // get a copy of the arg pointers as [Application new] removes
  // -NXHost options
  for(i=0; i<argc; i++)
    myArgv[i]=argv[i];
  
  [Application new];
  appName=[NXApp appName];
					       
  if(strcmp(appName, "open"))
    if(!strcmp(appName, "appopen") || !strcmp(appName, "run"))
    {
      if(argc<2)
      {
	printf("Usage: %s Application file...\n", appName);
	exit(1);
      }
      openApp=myArgv[arg++];
    } else if(!strcmp(appName, "unhide")) {
      if(argc>1)
	openApp=myArgv[arg++];
      openMode=UNHIDE;
    } else
      openApp=appName;
  
  while(1)
  {
    int argIsFile=0;
    
    if(arg>=argc)
      /* auto detect stdin */
      if(openedAFile)
	break;
      else {
	if((openApp!=WORKSPACE && openMode==OPEN) || openMode==UNHIDE)
	{
	  if(openMode==OPEN)
	    openMode=APP;
	  fileMode=NONE;
	} else
	  fileMode=STDIN;
	argIsFile++;
      }
    else if(myArgv[arg][0]=='-') {
      if(!strcmp(&myArgv[arg][1], "NXHost"))
      {
	if(arg!=(argc-1))
	  host=myArgv[++arg];
	else argIsFile++;
      } else if(!strcmp(&myArgv[arg][1], "wait")) {
	waitForFileChange++;
      } else if(!strcmp(&myArgv[arg][1], "nostat")) {
	noStat++;
      } else if(!strcmp(&myArgv[arg][1], "unhide") ||
		!strcmp(&myArgv[arg][1], "activate") ||
		!strcmp(&myArgv[arg][1], "makeKey")) {
	openMode=UNHIDE;
	fileMode=NONE;
	argIsFile++;
      } else if(!strcmp(&myArgv[arg][1], "temp")) {
	fileMode=TEMP;
      } else if(myArgv[arg][1]=='\000') {
	if(arg==(argc-1) && !openedAFile)
	  fileMode=STDIN;
	argIsFile++;
      } else if(myArgv[arg][2]=='\000') {
	switch(myArgv[arg][1]) {
	case 'a': /* next arg is new application to open with */
	  if(arg!=(argc-1))
	    openApp=myArgv[++arg];
	  else argIsFile++;
	  break;
	case 'o': /* open next arg without checking switches */
	  openMode=OPEN;
	  if(arg!=(argc-1))
	    arg++;
	  argIsFile++;
	  break;
	case 'p': /* go printmode */
	  openMode=PRINT;
	  break;
	default: /* not a switch -> open it */
	  argIsFile++;
	  break;
	}
      } else argIsFile++; /* not a longopt -> open it */
    } else if(myArgv[arg][0]=='+') {
      int i=0;
      
      oline=0;
      while(myArgv[arg][++i]!='\0')
	if(isdigit(myArgv[arg][i]))
	  oline=10*oline+(myArgv[arg][i]-48);
	else {
	  argIsFile++;
	  break;
	}
      if(!argIsFile)
      {
	if(arg!=(argc-1))
	{
	  openMode=OPENLINE;
	  arg++;
	}
	argIsFile++;
      }
    } else argIsFile++; /* not an opt at all -> open it */

    if(argIsFile)
    {
      char fname[MAXPATHLEN+1+(arg>=argc?0:strlen(myArgv[arg]))+1];
      struct stat buf;
      
      switch(fileMode) {
      case STDIN: {
	char buffer[STDINBUFSIZE+1];
	int len;
	FILE *file;
	
	if((len=fread(buffer, 1, STDINBUFSIZE, stdin))>=0)
	{
	  buffer[len]='\0';
	  if(len>=5 && !strncmp(buffer, "{\\rtf", 5))
	    strcpy(fname, "/tmp/openfiletmpXXXXXX.rtf");
	  else if(len>=11 && !strncmp(buffer, "%!PS-Adobe-", 11))
	    if(strstr(buffer, "EPSF-")<strchr(buffer, '\012'))
	      strcpy(fname, "/tmp/openfiletmpXXXXXX.eps");
	    else
	      strcpy(fname, "/tmp/openfiletmpXXXXXX.ps");
	  else
	    strcpy(fname, "/tmp/openfiletmpXXXXXX");
	  
	  NXGetTempFilename(fname, 16);
	  if(!(file=fopen(fname, "wb")))
	  {
	    printf("%s: couldn't create temp file\n", appName);
	    exit(1);
	  }
	  while(len==STDINBUFSIZE)
	  {
	    fwrite(buffer, STDINBUFSIZE, 1, file);
	    len=fread(buffer, 1, STDINBUFSIZE, stdin);
	  }
	  fwrite(buffer, 1, len, file);
	  fclose(file);
	}
	break;
      }
	case NONE:
	  break;
	case DISK:
	case TEMP:
	default:
	  if(myArgv[arg][0]=='/')
	    strcpy(fname, myArgv[arg]);
	  else {
	    getwd(fname);
	    strcat(fname, "/");
	    strcat(fname, myArgv[arg]);
	  }
	  if(fname[strlen(fname)-1]=='/')
	    strcat(fname, ".");
	  break;
	}

      if(!noStat && stat(fname, &buf))
	printf("%s: can't stat file: %s\n", appName, fname);
      else {
	char *connName;

	if((openMode==PRINT || openMode==OPENLINE) &&
	   !strcmp(openApp, WORKSPACE))
	  [[Application workspace] getInfoForFile:fname
	   application:&connName type:&NXPlainFileType];
	else
	  connName=str_copy(openApp);
	if((appPort=[appList getPortFor:connName onHost:host]) !=
	   PORT_NULL)
	{
	  [appSpeaker setSendPort:appPort];
	  switch(openMode) {
	  case UNHIDE:
	    [appSpeaker performRemoteMethod:"unhide"];
	    break;
	  case APP:
	    break;
	    
	  case OPENLINE:
	    [appSpeaker selectorRPC:"openFile:onHost:atTrueLine:"
	     paramTypes:"cci", fname, host?host:"localhost", oline];
	    break;
	  case PRINT:
	    if([appSpeaker msgPrint:fname ok:&ok] || !ok)
	      printf("%s: unable to print file: %s\n", appName, fname);
	    break;
	  case OPEN:
	  default:
	    if(fileMode==STDIN || fileMode==TEMP)
	    {
	      if([appSpeaker openTempFile:fname ok:&ok] || !ok)
		printf("%s: unable to open temp file: %s\n", appName,
		       fname);
	    } else {
	      if([appSpeaker openFile:fname ok:&ok] || !ok)
		printf("%s: unable to open file: %s\n", appName,
		       fname);
	    }
	    break;
	  }
	  if(!noStat && waitForFileChange)
	  {
	    time_t last=buf.st_mtime;
	    
	    while(buf.st_mtime==last && !sleep(1))
	      stat(fname, &buf);
	  }
	} else {
	  printf("%s: can't open connection to %s on %s.\n",
		 appName , openApp, host?host:"local host");
	}
	str_free(connName);
      }
      waitForFileChange=0;
      noStat=0;
      fileMode=DISK;
      openMode=OPEN;
      openedAFile++;
    }
    arg++;
  }
  
  [appSpeaker free];
  exit(0);
}
